RFC: Ginkgo: First bootstrap of using BeforeAll and AfterAll

eloycoto opened the pull request:
This commit fixed our outstanding issue about creating and deleting containers
for the same Testcase and do not have the option for setup and teardown
actions.

This code added a new vendor dependency, "ginkgo-ext" that it's a copy
of `Mirantis/Virtlet` e2e ginkgo-extension with a few additions to
support `--focus` option.

The main changes in this code are:

- `BeforeAll()` replaces BeforeSuite. On the other hand, `BeforeAll` can
be used on Describe and Context  like the initialize function that we're
using now.

- `AfterAll()` replaces `AfterSuite()`. On the other hand, `AfterAll()`
can be used in Describe and Context to clean the state after the
testCase finished, like tearDown function in any unit testing function.

The example TestCase can be the following:

```
var _ = Describe("K8Spolicies", func() {

        AfterAll(func() {
                fmt.Println("AfterAll---<")
        })

		AfterEach(func(){
			fmt.Println("AfterEach")
		})

        Context("Test Group", func() {

                AfterAll(func() {
                        fmt.Println("AfterAll---<Context")
                })

                It("Test1", func() {
                        Expect(true).To(BeTrue())
                })

                It("Test2", func() {
                        Expect(true).To(BeTrue())
                })
        })

        It("Test1", func() {
                Expect(true).To(BeTrue())
        })

        It("Test2", func() {
                Expect(true).To(BeTrue())
        })
})
```

ATM I only migrated `RuntimeValidatedMonitor`, where the AfterAll/BeforeEach is
used, the idea is to run the test for a few days, and if nothing fails move to
other testCases.

Pending things: 

- [] Move eloycoto/ginkgo-ext to cilium org.

eloycoto committed:
Ginkgo: First bootstrap of using BeforeAll and AfterAll

This commit fixed our outstanding issue about creating and deleting containers
for the same Testcase and do not have the option for setup and teardown
actions.

This code added a new vendor dependency, "ginkgo-ext" that it's a copy
of `Mirantis/Virtlet` e2e ginkgo-extension with a few additions to
support `--focus` option.

The main changes in this code are:

- `BeforeAll()` replaces BeforeSuite. On the other hand, `BeforeAll` can
be used on Describe and Context  like the initialize function that we're
using now.

- `AfterAll()` replaces `AfterSuite()`. On the other hand, `AfterAll()`
can be used in Describe and Context to clean the state after the
testCase finished, like tearDown function in any unit testing function.

The example TestCase can be the following:

```
var _ = Describe("K8Spolicies", func() {

        AfterAll(func() {
                fmt.Println("AfterAll---<")
        })

		AfterEach(func(){
			fmt.Println("AfterEach")
		})

        Context("Test Group", func() {

                AfterAll(func() {
                        fmt.Println("AfterAll---<Context")
                })

                It("Test1", func() {
                        Expect(true).To(BeTrue())
                })

                It("Test2", func() {
                        Expect(true).To(BeTrue())
                })
        })

        It("Test1", func() {
                Expect(true).To(BeTrue())
        })

        It("Test2", func() {
                Expect(true).To(BeTrue())
        })
})
```

ATM I only migrated `RuntimeValidatedMonitor`, where the AfterAll/BeforeEach is
used, the idea is to run the test for a few days, and if nothing fails move to
other testCases.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

aanm reviewed (commented):
This looks sane to me but why not simply copy the `vendor/github.com/eloycoto/ginkgo-ext/scopes.go` to `/test` directory?

ianvernon commented on test/test_suite_test.go:
```diff
@@ -78,8 +79,10 @@ func TestTest(t *testing.T) {
+	// RunSpecsWithDefaultAndCustomReporters(
+	// 	t, ginkgoext.GetScopeWithVersion(), []ginkgo.Reporter{junitReporter})
+
+	RunSpecsWithDefaultAndCustomReporters(t, "Foo Suite", []ginkgo.Reporter{junitReporter})
```
What is the need for this change?

eloycoto replied:
leftover

eloycoto committed:
Ginkgo: Importing mirantis/Virtlet code

This code was originally created by Stan Lagun
(https://github.com/istalker2) in the project
https://github.com/Mirantis/virtlet

This code is a copy and paste of
https://github.com/istalker2/virtlet/blob/master/tests/e2e/ginkgo-ext/scopes.go

eloycoto committed:
Ginkgo: Added --focus support in BeforeAll/AfterAll

Update ginkgo-ext dependency
Deleted ginkgo-ext dependency

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>
